### PR TITLE
[INTEL oneDNN] user mode scratchpad implementation for inner-product

### PR DESCRIPTION
**Pull request by gzmkl**
Enable user mode scratchpad for inner-product (matmul fused & quantized matmul) for better memory usage control.

Note: code change with mkl_util.h is NOT needed once PR

  https://github.com/tensorflow/tensorflow/pull/55256

is merged.

### CODE REVIEW

**Comment by gzmkl** [comment]
@penpornk  The new code in mkl_util.h causes Windows compilation failure (same code in the recently reverted PR).
We will fix it.

So please hold on the review.  Thanks!

**Comment by gzmkl** [comment]
@penpornk  This PR is ready for review. 

You can combine it with a related scratchpad PR 

https://github.com/tensorflow/tensorflow/pull/55256

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR!

File: tensorflow/core/util/mkl_util.h [hunk 1]
```diff
@@ -802,6 +802,38 @@ inline void AllocTmpBuffer(OpKernelContext* context, Tensor* tensor_out,
...
+  //      spad_md.get_size() will return 0. It is fine to return
+  //      nullptr in this case
+  inline void AllocateSPTensor(MklPrim* mkl_prim, OpKernelContext* context) {
+    allocated_ = false;
+    auto spad_md = mkl_prim->GetScratchPadDesc();
+    size_t spad_size = spad_md.get_size();
+    if (spad_size == 0) return;
+
+    size_t allocate_size =
+        (spad_size / sizeof(T)) + ((spad_size % sizeof(T) > 0) ? 1 : 0);
```
[the remark this answers is not in the document]
Comment (gzmkl) [reply tensorflow/core/util/mkl_util.h]: @penpornk  Made code change per suggestion. Thanks!

I also committed the same change to my other scratchpad PR (for conv op).

**Comment by gzmkl** [comment]
Done with change

**Review by penpornk — APPROVED** [review]
Thank you for the change!

**Comment by penpornk** [comment]
@gzmkl Could you please help resolve merge conflict? Thank you very much!

**Review by penpornk — APPROVED** [review]
Thank you!

**Comment by penpornk** [comment]
Manually close this PR since it is merged in https://github.com/tensorflow/tensorflow/commit/2ed1ca2e69a3200f70d7025f260e5d97f82b18b9.

**Pull request closed without merging**